win32ss/user/ntuser/ghost.c

```diff
@@ -62,89 +120,297 @@ BOOL FASTCALL IntIsGhostWindow(PWND Window)
+    pGhostWnd = co_UserCreateWindowEx(&Cs, &GhostClass, &WindowName, NULL);
+    if (!pGhostWnd)
+    {
+        ERR("Failed to create ghost\n");
+        goto Quit;
+    }
+
+    // add to list
+    bRet = IntAddGhostEntry(hwndTarget, pGhostWnd->head.h);
+
+Quit:
+    if (!bRet && pGhostWnd)
+    {
+        // destroy ghost
+        co_UserDestroyWindow(pGhostWnd);
+    }
+
+    // resume hooks
+    if (!NoHooks)
+    {
+        ptiCurrent->TIF_flags &= ~TIF_DISABLEHOOKS;
+        ptiCurrent->pClientInfo->dwTIFlags = ptiCurrent->TIF_flags;
+    }
+
+    return bRet;
 }
 
-HWND FASTCALL IntHungWindowFromGhostWindow(PWND pGhostWnd)
+BOOL FASTCALL
+IntGhostOnDestroy(HWND hwndGhost, HWND hwndTarget)
 {
-    const GHOST_DATA *UserData;
-    HWND hwndTarget;
+    PWND pGhostWnd;
+
+    pGhostWnd = ValidateHwndNoErr(hwndGhost);
+    if (!pGhostWnd)
+    {
+        ERR("Not a window.\n");
+        return FALSE;
+    }
 
     if (!IntIsGhostWindow(pGhostWnd))
     {
-        DPRINT("Not a ghost window\n");
-        return NULL;
+        ERR("Not a Ghost window.\n");
+        return FALSE;
     }
 
-    UserData = (const GHOST_DATA *)pGhostWnd->dwUserData;
-    if (UserData)
+    return IntDeleteGhostEntry(hwndTarget);
+}
+
+/*
+ * GhostThreadMain
+ *
+ * Creates ghost windows and exits when no non-responsive window remains.
+ */
+VOID NTAPI
+UserGhostThreadEntry(VOID)
+{
+    // This thread should handle all ghost windows and exit when no ghost window is needed.
+    BOOL bInitialized = FALSE;
+    MSG msg;
+    HDESK hDesk;
+
+    TRACE("UserGhostThreadEntry: started.\n");
+
+    ASSERT(!UserIsEntered());
+    UserEnterExclusive();
+
+    // attach thread to desktop
+    hDesk = IntGetDesktopObjectHandle(gpdeskInputDesktop);
+    IntSetThreadDesktop(hDesk, FALSE);
+
+    // get a PTHREADINFO
+    gptiGhostThread = PsGetCurrentThreadWin32Thread();
+
+    // thread message loop
+    for (;;)
     {
-        _SEH2_TRY
+        RtlZeroMemory(&msg, sizeof(MSG));
+
+        if (bInitialized && IsListEmpty(&gGhostListHead))
+        {
+            break;
+        }
+
+        if (!co_IntGetPeekMessage(&msg, NULL, 0, 0, PM_REMOVE, TRUE))
         {
-            ProbeForRead(UserData, sizeof(GHOST_DATA), 1);
-            hwndTarget = UserData->hwndTarget;
+            InitializeListHead(&gGhostListHead);
+            break;
         }
-        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+
+        switch (msg.message)
         {
-            DPRINT1("Exception!\n");
-            hwndTarget = NULL;
+            case GTM_CREATE_GHOST:
+            {
+                HWND hwndTarget = (HWND)msg.wParam;     // not trusted yet
+
+                TRACE("GTM_CREATE_GHOST: %p\n", (void *)hwndTarget);
+
+                if (IntCreateGhost(hwndTarget))
+                {
+                    TRACE("IntCreateGhost: Success.\n");
+                }
+
+                // pulse event
+                KePulseEvent(gpGhostStartupEvent, IO_NO_INCREMENT, FALSE);
```

Comment:
I'm not sure I've ever seen a legitimate use of KePulseEvent.

If you want my recommendation, here's how I would synchronize the threads:

Ghost thread:
```
UserEnter
gptiGhostThread = PsGetCurrentThreadWin32Thread()
KeSetEvent(StartupEvent)
while (!IsListEmpty(gGhostList))
{
    // process messages
}
gptiGhostThread = NULL;
UserLeave
```

Other thread:
```
UserEnter
AddStuffTo(gGhostList)
if (!gptiGhostThread)
{
    StartupEvent = CreateEvent()
    CreateThread()
    UserLeave()
    KeWait(StartupEvent)
    UserEnter
    DeleteEvent(StartupEvent)
}
PostThreadMessage
```


Another way to structure this (which is closer to what Windows does) would be this, where work requests are saved in the lists and there's an event that indicates there's work to be done (rather than posting messages).
Ghost thread:
```
UserEnter
gptiGhostThread = PsGetCurrentThreadWin32Thread()
while (!IsListEmpty(gGhostThreadList))
{
    Status = MsgWaitForMultipleObjects(... GhostThreadHasWorkEvent ...)
    if (Status == STATUS_WAIT_0)
    {
        WalkListAndProcess(gGhostList)
    }
    else
    {
        // process message
    }

}
gptiGhostThread = NULL
DeleteEvent(GhostThreadHasWorkEvent)
UserLeave
```

Other thread:
```
UserEnter
AddStuffTo(gGhostList)
if (!gptiGhostThread)
{
    GhostThreadHasWorkEvent = CreateEvent()
    CreateThread()
}
KeSetEvent(GhostThreadHasWorkEvent)
```

Reply:
Chose 1st option.